{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-03-18 01:15:49.937379: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2025-03-18 01:15:50.052754: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2025-03-18 01:15:50.088427: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2025-03-18 01:15:50.387467: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2025-03-18 01:15:52.119402: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "import sys\n",
    "import os\n",
    "import glob\n",
    "import math\n",
    "import threading\n",
    "import concurrent.futures as cf\n",
    "import random\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tensorflow as tf\n",
    "from keras import Input, Model, layers, metrics, losses, callbacks, optimizers, models, utils\n",
    "from keras import backend as K\n",
    "import gc\n",
    "import keras_tuner as kt\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "K.clear_session()\n",
    "gc.collect()\n",
    "\n",
    "datasets_path = \"../../Datasets/\"\n",
    "models_path = \"../../Models/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook was used to calculate intron boundaries using the hg38 basic annotations gtf."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_gtf_annotations(gtf_file):\n",
    "    \"\"\"\n",
    "    Loads GTF into a pandas DataFrame and converts cstart and cend to zero-based indexing.\n",
    "    \"\"\"\n",
    "    gtf_data = pd.read_csv(\n",
    "        gtf_file, sep='\\t', comment='#', header=None,\n",
    "        names=['seqname', 'source', 'feature', 'cstart', 'cend', \n",
    "               'score', 'strand', 'frame', 'attribute']\n",
    "    )\n",
    "    # Convert to zero-based indexing for cstart\n",
    "    gtf_data['cstart'] = gtf_data['cstart'] - 1\n",
    "    return gtf_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11120</td>\n",
       "      <td>24894</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.2\"; gene_type \"lncRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11425</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11425</td>\n",
       "      <td>11671</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source     feature  cstart   cend score strand frame  \\\n",
       "0    chr1  HAVANA        gene   11120  24894     .      +     .   \n",
       "1    chr1  HAVANA  transcript   11425  14409     .      +     .   \n",
       "2    chr1  HAVANA        exon   11425  11671     .      +     .   \n",
       "3    chr1  HAVANA        exon   12009  12227     .      +     .   \n",
       "4    chr1  HAVANA        exon   12612  12721     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000290825.2\"; gene_type \"lncRNA...  \n",
       "1  gene_id \"ENSG00000290825.2\"; transcript_id \"EN...  \n",
       "2  gene_id \"ENSG00000290825.2\"; transcript_id \"EN...  \n",
       "3  gene_id \"ENSG00000290825.2\"; transcript_id \"EN...  \n",
       "4  gene_id \"ENSG00000290825.2\"; transcript_id \"EN...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_data = load_gtf_annotations(datasets_path + 'basic_annotations.gtf')\n",
    "# annotation_data = annotation_data[annotation_data[\"seqname\"]!=\"chrM\"]\n",
    "annotation_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25\n",
      "['chr1' 'chr2' 'chr3' 'chr4' 'chr5' 'chr6' 'chr7' 'chr8' 'chr9' 'chr10'\n",
      " 'chr11' 'chr12' 'chr13' 'chr14' 'chr15' 'chr16' 'chr17' 'chr18' 'chr19'\n",
      " 'chr20' 'chr21' 'chr22' 'chrX' 'chrY' 'chrM']\n"
     ]
    }
   ],
   "source": [
    "seqnames = annotation_data[\"seqname\"]\n",
    "unique_features = seqnames.unique()\n",
    "print(len(unique_features))\n",
    "print(unique_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = annotation_data[annotation_data[\"feature\"]==\"exon\"]\n",
    "genes = annotation_data[annotation_data[\"feature\"]==\"gene\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_exons = exons[exons['seqname']!='chrM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "63049\n"
     ]
    }
   ],
   "source": [
    "rows, columns = genes.shape\n",
    "print(rows) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.6\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.6\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  source feature  cstart   cend score strand frame  \\\n",
       "2    chr1  HAVANA    exon   11868  12227     .      +     .   \n",
       "3    chr1  HAVANA    exon   12612  12721     .      +     .   \n",
       "4    chr1  HAVANA    exon   13220  14409     .      +     .   \n",
       "7    chr1  HAVANA    exon   12009  12057     .      +     .   \n",
       "8    chr1  HAVANA    exon   12178  12227     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "2  gene_id \"ENSG00000290825.1\"; transcript_id \"EN...  \n",
       "3  gene_id \"ENSG00000290825.1\"; transcript_id \"EN...  \n",
       "4  gene_id \"ENSG00000290825.1\"; transcript_id \"EN...  \n",
       "7  gene_id \"ENSG00000223972.6\"; transcript_id \"EN...  \n",
       "8  gene_id \"ENSG00000223972.6\"; transcript_id \"EN...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.668590e+06</td>\n",
       "      <td>1.668590e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.616518e+07</td>\n",
       "      <td>7.616545e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.626089e+07</td>\n",
       "      <td>5.626089e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.110200e+04</td>\n",
       "      <td>1.159500e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.254061e+07</td>\n",
       "      <td>3.254121e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.382113e+07</td>\n",
       "      <td>6.382138e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.119674e+08</td>\n",
       "      <td>1.119675e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489366e+08</td>\n",
       "      <td>2.489370e+08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cstart          cend\n",
       "count  1.668590e+06  1.668590e+06\n",
       "mean   7.616518e+07  7.616545e+07\n",
       "std    5.626089e+07  5.626089e+07\n",
       "min    1.110200e+04  1.159500e+04\n",
       "25%    3.254061e+07  3.254121e+07\n",
       "50%    6.382113e+07  6.382138e+07\n",
       "75%    1.119674e+08  1.119675e+08\n",
       "max    2.489366e+08  2.489370e+08"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "shared_starts = exons.groupby(['seqname', 'feature', 'cstart', 'strand']).size().reset_index(name='count')\n",
    "shared_ends = exons.groupby(['seqname', 'feature', 'cend', 'strand']).size().reset_index(name='count')\n",
    "shared_both = exons.groupby(['seqname', 'feature', 'cstart', 'cend', 'strand']).size().reset_index(name='count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cstart</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>5.221340e+05</td>\n",
       "      <td>522134.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.485583e+07</td>\n",
       "      <td>3.195712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.596011e+07</td>\n",
       "      <td>5.246876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.110200e+04</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.141290e+07</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.186045e+07</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.107137e+08</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489366e+08</td>\n",
       "      <td>229.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cstart          count\n",
       "count  5.221340e+05  522134.000000\n",
       "mean   7.485583e+07       3.195712\n",
       "std    5.596011e+07       5.246876\n",
       "min    1.110200e+04       1.000000\n",
       "25%    3.141290e+07       1.000000\n",
       "50%    6.186045e+07       1.000000\n",
       "75%    1.107137e+08       3.000000\n",
       "max    2.489366e+08     229.000000"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared_starts.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cend</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>5.216560e+05</td>\n",
       "      <td>521656.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.488912e+07</td>\n",
       "      <td>3.198640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.593222e+07</td>\n",
       "      <td>5.252831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.159500e+04</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.142690e+07</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.192027e+07</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.107278e+08</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489370e+08</td>\n",
       "      <td>235.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               cend          count\n",
       "count  5.216560e+05  521656.000000\n",
       "mean   7.488912e+07       3.198640\n",
       "std    5.593222e+07       5.252831\n",
       "min    1.159500e+04       1.000000\n",
       "25%    3.142690e+07       1.000000\n",
       "50%    6.192027e+07       1.000000\n",
       "75%    1.107278e+08       3.000000\n",
       "max    2.489370e+08     235.000000"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared_ends.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>6.703820e+05</td>\n",
       "      <td>6.703820e+05</td>\n",
       "      <td>670382.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.468505e+07</td>\n",
       "      <td>7.468544e+07</td>\n",
       "      <td>2.489014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.587040e+07</td>\n",
       "      <td>5.587041e+07</td>\n",
       "      <td>4.339533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.110200e+04</td>\n",
       "      <td>1.159500e+04</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.147350e+07</td>\n",
       "      <td>3.147375e+07</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.163945e+07</td>\n",
       "      <td>6.164110e+07</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.104317e+08</td>\n",
       "      <td>1.104320e+08</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489366e+08</td>\n",
       "      <td>2.489370e+08</td>\n",
       "      <td>215.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cstart          cend          count\n",
       "count  6.703820e+05  6.703820e+05  670382.000000\n",
       "mean   7.468505e+07  7.468544e+07       2.489014\n",
       "std    5.587040e+07  5.587041e+07       4.339533\n",
       "min    1.110200e+04  1.159500e+04       1.000000\n",
       "25%    3.147350e+07  3.147375e+07       1.000000\n",
       "50%    6.163945e+07  6.164110e+07       1.000000\n",
       "75%    1.104317e+08  1.104320e+08       2.000000\n",
       "max    2.489366e+08  2.489370e+08     215.000000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared_both.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1668590\n"
     ]
    }
   ],
   "source": [
    "rows, columns = exons.shape\n",
    "print(rows) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1668590\n"
     ]
    }
   ],
   "source": [
    "rows, columns = chr_exons.shape\n",
    "print(rows) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "37 genes and thus 37 exons on chrM removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.1\"; gene_type \"lncRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>12009</td>\n",
       "      <td>13670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.6\"; gene_type \"transc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14695</td>\n",
       "      <td>24886</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227232.6\"; gene_type \"unproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17368</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278267.1\"; gene_type \"miRNA\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>29553</td>\n",
       "      <td>31109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243485.5\"; gene_type \"lncRNA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467013</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>576</td>\n",
       "      <td>647</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210049.1\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467016</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>647</td>\n",
       "      <td>1601</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000211459.2\"; gene_type \"Mt_rRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467019</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>1601</td>\n",
       "      <td>1670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210077.1\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467022</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>1670</td>\n",
       "      <td>3229</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210082.2\"; gene_type \"Mt_rRN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467025</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3229</td>\n",
       "      <td>3304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000209082.1\"; gene_type \"Mt_tRN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>125 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source feature  cstart   cend score strand frame  \\\n",
       "0          chr1   HAVANA    gene   11868  14409     .      +     .   \n",
       "5          chr1   HAVANA    gene   12009  13670     .      +     .   \n",
       "13         chr1   HAVANA    gene   14695  24886     .      -     .   \n",
       "25         chr1  ENSEMBL    gene   17368  17436     .      -     .   \n",
       "28         chr1   HAVANA    gene   29553  31109     .      +     .   \n",
       "...         ...      ...     ...     ...    ...   ...    ...   ...   \n",
       "3467013    chrM  ENSEMBL    gene     576    647     .      +     .   \n",
       "3467016    chrM  ENSEMBL    gene     647   1601     .      +     .   \n",
       "3467019    chrM  ENSEMBL    gene    1601   1670     .      +     .   \n",
       "3467022    chrM  ENSEMBL    gene    1670   3229     .      +     .   \n",
       "3467025    chrM  ENSEMBL    gene    3229   3304     .      +     .   \n",
       "\n",
       "                                                 attribute  \n",
       "0        gene_id \"ENSG00000290825.1\"; gene_type \"lncRNA...  \n",
       "5        gene_id \"ENSG00000223972.6\"; gene_type \"transc...  \n",
       "13       gene_id \"ENSG00000227232.6\"; gene_type \"unproc...  \n",
       "25       gene_id \"ENSG00000278267.1\"; gene_type \"miRNA\"...  \n",
       "28       gene_id \"ENSG00000243485.5\"; gene_type \"lncRNA...  \n",
       "...                                                    ...  \n",
       "3467013  gene_id \"ENSG00000210049.1\"; gene_type \"Mt_tRN...  \n",
       "3467016  gene_id \"ENSG00000211459.2\"; gene_type \"Mt_rRN...  \n",
       "3467019  gene_id \"ENSG00000210077.1\"; gene_type \"Mt_tRN...  \n",
       "3467022  gene_id \"ENSG00000210082.2\"; gene_type \"Mt_rRN...  \n",
       "3467025  gene_id \"ENSG00000209082.1\"; gene_type \"Mt_tRN...  \n",
       "\n",
       "[125 rows x 9 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000290825.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12009</td>\n",
       "      <td>12057</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.6\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12178</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.6\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467015</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>576</td>\n",
       "      <td>647</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210049.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467018</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>647</td>\n",
       "      <td>1601</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000211459.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467021</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>1601</td>\n",
       "      <td>1670</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210077.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467024</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>1670</td>\n",
       "      <td>3229</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210082.2\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3467027</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>3229</td>\n",
       "      <td>3304</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000209082.1\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>125 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source feature  cstart   cend score strand frame  \\\n",
       "2          chr1   HAVANA    exon   11868  12227     .      +     .   \n",
       "3          chr1   HAVANA    exon   12612  12721     .      +     .   \n",
       "4          chr1   HAVANA    exon   13220  14409     .      +     .   \n",
       "7          chr1   HAVANA    exon   12009  12057     .      +     .   \n",
       "8          chr1   HAVANA    exon   12178  12227     .      +     .   \n",
       "...         ...      ...     ...     ...    ...   ...    ...   ...   \n",
       "3467015    chrM  ENSEMBL    exon     576    647     .      +     .   \n",
       "3467018    chrM  ENSEMBL    exon     647   1601     .      +     .   \n",
       "3467021    chrM  ENSEMBL    exon    1601   1670     .      +     .   \n",
       "3467024    chrM  ENSEMBL    exon    1670   3229     .      +     .   \n",
       "3467027    chrM  ENSEMBL    exon    3229   3304     .      +     .   \n",
       "\n",
       "                                                 attribute  \n",
       "2        gene_id \"ENSG00000290825.1\"; transcript_id \"EN...  \n",
       "3        gene_id \"ENSG00000290825.1\"; transcript_id \"EN...  \n",
       "4        gene_id \"ENSG00000290825.1\"; transcript_id \"EN...  \n",
       "7        gene_id \"ENSG00000223972.6\"; transcript_id \"EN...  \n",
       "8        gene_id \"ENSG00000223972.6\"; transcript_id \"EN...  \n",
       "...                                                    ...  \n",
       "3467015  gene_id \"ENSG00000210049.1\"; transcript_id \"EN...  \n",
       "3467018  gene_id \"ENSG00000211459.2\"; transcript_id \"EN...  \n",
       "3467021  gene_id \"ENSG00000210077.1\"; transcript_id \"EN...  \n",
       "3467024  gene_id \"ENSG00000210082.2\"; transcript_id \"EN...  \n",
       "3467027  gene_id \"ENSG00000209082.1\"; transcript_id \"EN...  \n",
       "\n",
       "[125 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>3.467013e+06</td>\n",
       "      <td>3.467013e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.589303e+07</td>\n",
       "      <td>7.589686e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.615757e+07</td>\n",
       "      <td>5.615783e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.110200e+04</td>\n",
       "      <td>1.159500e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.260765e+07</td>\n",
       "      <td>3.261128e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.345127e+07</td>\n",
       "      <td>6.346637e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.116243e+08</td>\n",
       "      <td>1.116324e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489366e+08</td>\n",
       "      <td>2.489370e+08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cstart          cend\n",
       "count  3.467013e+06  3.467013e+06\n",
       "mean   7.589303e+07  7.589686e+07\n",
       "std    5.615757e+07  5.615783e+07\n",
       "min    1.110200e+04  1.159500e+04\n",
       "25%    3.260765e+07  3.261128e+07\n",
       "50%    6.345127e+07  6.346637e+07\n",
       "75%    1.116243e+08  1.116324e+08\n",
       "max    2.489366e+08  2.489370e+08"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_data.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "features = annotation_data[\"feature\"]\n",
    "unique_features = features.unique()\n",
    "print(len(unique_features))\n",
    "print(unique_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_introns(gtf_df):\n",
    "    \"\"\"\n",
    "    Given a pandas DataFrame of GTF records (with columns: 'seqname', 'feature', 'start',\n",
    "    'end', 'strand', and 'attribute') in which the start (called cstart) has been converted\n",
    "    to 0-indexed values, calculate the introns for each gene.\n",
    "    \n",
    "    This function assumes that the attribute field contains a gene identifier in a form like:\n",
    "       gene_id \"XYZ\";\n",
    "    as it does in the hg38 gtf and groups features by gene_id. For each gene, it collects the exon \n",
    "    intervals, merges overlapping exons (forms the union of exonic regions) and then computes each intron \n",
    "    as the gap between consecutive merged exons. For plus-strand genes the intron is reported\n",
    "    as (previous_exon.end, next_exon.start), while for minus-strand genes the order is reversed so\n",
    "    that the cstart value is higher than the cend.\n",
    "    \n",
    "    Returns:\n",
    "        A new DataFrame with one row per intron, having columns:\n",
    "          - seqname\n",
    "          - feature (with value \"intron\")\n",
    "          - cstart (the start coordinate in 0-index system; note that for minus strand this is numerically higher)\n",
    "          - cend   (the end coordinate)\n",
    "          - strand\n",
    "    \"\"\"\n",
    "    \n",
    "    # helper to extract gene_id from the attribute string\n",
    "    def get_gene_id(attr):\n",
    "        # look for a pattern like: gene_id \"XYZ\";\n",
    "        m = re.search(r'gene_id\\s+\"([^\"]+)\"', attr)\n",
    "        if m:\n",
    "            return m.group(1)\n",
    "        else:\n",
    "            return None\n",
    "\n",
    "    # Add a gene_id column (if not already present)\n",
    "    if 'gene_id' not in gtf_df.columns:\n",
    "        gtf_df = gtf_df.copy()  # avoid modifying the original dataframe\n",
    "        gtf_df['gene_id'] = gtf_df['attribute'].apply(get_gene_id)\n",
    "    \n",
    "    intron_records = []\n",
    "    \n",
    "    # Group rows by gene_id (each gene should have one gene-level record, and one or more exon records)\n",
    "    for gene_id, group in gtf_df.groupby('gene_id'):\n",
    "        # Skip groups with no gene_id (if any)\n",
    "        if gene_id is None:\n",
    "            continue\n",
    "        \n",
    "        # Identify the gene-level information (if available)\n",
    "        gene_rows = group[group['feature'] == 'gene']\n",
    "        if not gene_rows.empty:\n",
    "            # Use the gene row to get the chromosome and strand.\n",
    "            seqname = gene_rows.iloc[0]['seqname']\n",
    "            strand  = gene_rows.iloc[0]['strand']\n",
    "            gene_start = gene_rows.iloc[0]['cstart']\n",
    "            gene_end   = gene_rows.iloc[0]['cend']\n",
    "        else:\n",
    "            # Fall back on the first exon if no gene record is available.\n",
    "            seqname = group.iloc[0]['seqname']\n",
    "            strand  = group.iloc[0]['strand']\n",
    "            gene_start = None\n",
    "            gene_end = None\n",
    "        \n",
    "        # Get all exon rows for this gene\n",
    "        exon_rows = group[group['feature'] == 'exon']\n",
    "        if exon_rows.empty:\n",
    "            continue\n",
    "        \n",
    "        # Build a list of exon intervals (each as a tuple (start, end))\n",
    "        # Optionally we could filter to exons that fall within the gene boundaries.\n",
    "        exon_intervals = list(zip(exon_rows['cstart'], exon_rows['cend']))\n",
    "        \n",
    "        # Sort by start (genomic order)\n",
    "        exon_intervals = sorted(exon_intervals, key=lambda x: x[0])\n",
    "        \n",
    "        # Merge overlapping or adjacent exons.\n",
    "        # (For example, if two exons overlap because of alternative splicing, we want the union.)\n",
    "        merged_exons = []\n",
    "        for interval in exon_intervals:\n",
    "            if not merged_exons:\n",
    "                merged_exons.append(list(interval))\n",
    "            else:\n",
    "                last = merged_exons[-1]\n",
    "                # If the current exon overlaps or touches the previous one, merge them.\n",
    "                if interval[0] <= last[1]:\n",
    "                    last[1] = max(last[1], interval[1])\n",
    "                else:\n",
    "                    merged_exons.append(list(interval))\n",
    "        \n",
    "        # If there is fewer than two merged exons, then there is no intron.\n",
    "        if len(merged_exons) < 2:\n",
    "            continue\n",
    "        \n",
    "        # For plus-strand genes, process the merged exons in genomic (ascending) order.\n",
    "        if strand == '+':\n",
    "            # For each adjacent pair of merged exons, define an intron between them.\n",
    "            for i in range(len(merged_exons) - 1):\n",
    "                # The intron is the gap between the end of exon i and the start of exon i+1.\n",
    "                intron_start = merged_exons[i][1]\n",
    "                intron_end   = merged_exons[i+1][0]\n",
    "                # Only add if there is a gap.\n",
    "                if intron_end > intron_start:\n",
    "                    intron_records.append({\n",
    "                        'seqname': seqname,\n",
    "                        'feature': 'intron',\n",
    "                        'cstart': intron_start,\n",
    "                        'cend': intron_end,\n",
    "                        'strand': strand\n",
    "                    })\n",
    "        else:\n",
    "            # For minus-strand genes, the transcript order is reversed.\n",
    "            # Sort the merged exons in descending order by their start coordinate.\n",
    "            merged_exons_desc = sorted(merged_exons, key=lambda x: x[0], reverse=True)\n",
    "            for i in range(len(merged_exons_desc) - 1):\n",
    "                # In transcript order the first exon is the one with the higher coordinate.\n",
    "                # We want to report the intron so that cstart is higher than cend.\n",
    "                intron_start = merged_exons_desc[i][0]   # start of the upstream exon\n",
    "                intron_end   = merged_exons_desc[i+1][1]   # end of the downstream exon\n",
    "                if intron_start > intron_end:\n",
    "                    intron_records.append({\n",
    "                        'seqname': seqname,\n",
    "                        'feature': 'intron',\n",
    "                        'cstart': intron_start,\n",
    "                        'cend': intron_end,\n",
    "                        'strand': strand\n",
    "                    })\n",
    "    \n",
    "    return pd.DataFrame(intron_records)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "324792\n"
     ]
    }
   ],
   "source": [
    "annotation_data = load_gtf_annotations(datasets_path + 'basic_annotations.gtf')\n",
    "annotation_data = annotation_data[annotation_data[\"seqname\"]!=\"chrM\"]\n",
    "introns = calculate_introns(annotation_data)\n",
    "print(len(introns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>feature</th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>intron</td>\n",
       "      <td>100636607</td>\n",
       "      <td>100635746</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>intron</td>\n",
       "      <td>100635557</td>\n",
       "      <td>100635252</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>intron</td>\n",
       "      <td>100635177</td>\n",
       "      <td>100634029</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>intron</td>\n",
       "      <td>100633930</td>\n",
       "      <td>100633539</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>intron</td>\n",
       "      <td>100633404</td>\n",
       "      <td>100632568</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname feature     cstart       cend strand\n",
       "0    chrX  intron  100636607  100635746      -\n",
       "1    chrX  intron  100635557  100635252      -\n",
       "2    chrX  intron  100635177  100634029      -\n",
       "3    chrX  intron  100633930  100633539      -\n",
       "4    chrX  intron  100633404  100632568      -"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "introns.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "trimmed_annotation_data = annotation_data[[\"seqname\", \"feature\", \"cstart\", \"cend\", \"strand\"]]\n",
    "IntronExonDF = pd.concat([trimmed_annotation_data, introns])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "IntronExonDF.to_csv(datasets_path + 'IntronExonDF.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "introns.to_csv(datasets_path + 'BetterIntrons.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def swap_columns_if_needed(df, col_a, col_b):\n",
    "    \"\"\"\n",
    "    Turns out the (-) strand lists cstart as smaller than cend.  This fixes the output from\n",
    "    the above function that calculated intron boundaries.\n",
    "    For each row in the dataframe, if the value in col_a is greater than the value in col_b,\n",
    "    swap the two values.\n",
    "\n",
    "    Parameters:\n",
    "        df (pd.DataFrame): The dataframe to process.\n",
    "        col_a (str): The name of the first column.\n",
    "        col_b (str): The name of the second column.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: The dataframe with swapped values where needed.\n",
    "    \"\"\"\n",
    "    # Create a boolean mask where the value in col_a is greater than col_b.\n",
    "    mask = df[col_a] > df[col_b]\n",
    "    \n",
    "    # Swap the values in col_a and col_b for rows where mask is True.\n",
    "    df.loc[mask, [col_a, col_b]] = df.loc[mask, [col_b, col_a]].values\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "FixedIntronExonDF = swap_columns_if_needed(IntronExonDF, 'cstart', 'cend')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['gene' 'transcript' 'exon' 'CDS' 'start_codon' 'stop_codon' 'UTR'\n",
      " 'Selenocysteine' 'intron']\n"
     ]
    }
   ],
   "source": [
    "print(FixedIntronExonDF[\"feature\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "Trimmed_Intron_Exon_DF = FixedIntronExonDF[((FixedIntronExonDF[\"feature\"]==\"exon\") | (FixedIntronExonDF[\"feature\"]==\"intron\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['exon' 'intron']\n"
     ]
    }
   ],
   "source": [
    "print(Trimmed_Intron_Exon_DF[\"feature\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "Trimmed_Intron_Exon_DF = Trimmed_Intron_Exon_DF[[\"seqname\", \"feature\", \"cstart\", \"cend\", \"strand\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        seqname feature     cstart       cend strand\n",
      "1627030   chr15    exon   88905756   88905901      -\n",
      "450582     chr3    exon   98849460   98849598      -\n",
      "1820139   chr17    exon   63862288   63862531      +\n",
      "208742     chr1    exon  241098665  241098762      -\n",
      "739233     chr6    exon   29945233   29945281      +\n",
      "139817    chr14  intron   88695454   88696846      -\n",
      "1185731   chr10    exon  112951482  112951607      +\n",
      "785077     chr6    exon   85536791   85536924      -\n",
      "270406     chr3  intron  170467402  170474264      +\n",
      "2194049    chrX    exon  119616711  119616742      -\n"
     ]
    }
   ],
   "source": [
    "print(Trimmed_Intron_Exon_DF.sample(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On human genome UCSC browser, due to indexing 1 on their end and 0 in python, cstart here is the last base to the right of the feature on the browser.\n",
    "cend looks like the correct spot, but only because python excludes the last base which cancels out the off by 1 issue\n",
    "\n",
    "Printing line by line swaps back into 1 indexing in a .txt file so locations are accurate as long as line 1 has the first base."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Trimmed_Intron_Exon_DF.to_csv(datasets_path + \"FinalIntronExonDF.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.007358e+06</td>\n",
       "      <td>1.007358e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.623405e+07</td>\n",
       "      <td>7.623434e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.645680e+07</td>\n",
       "      <td>5.645680e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>5.760000e+02</td>\n",
       "      <td>6.470000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.220527e+07</td>\n",
       "      <td>3.220547e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.381703e+07</td>\n",
       "      <td>6.381753e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.124123e+08</td>\n",
       "      <td>1.124125e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489367e+08</td>\n",
       "      <td>2.489371e+08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cstart          cend\n",
       "count  1.007358e+06  1.007358e+06\n",
       "mean   7.623405e+07  7.623434e+07\n",
       "std    5.645680e+07  5.645680e+07\n",
       "min    5.760000e+02  6.470000e+02\n",
       "25%    3.220527e+07  3.220547e+07\n",
       "50%    6.381703e+07  6.381753e+07\n",
       "75%    1.124123e+08  1.124125e+08\n",
       "max    2.489367e+08  2.489371e+08"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_exons = load_gtf_annotations(datasets_path + 'basic_annotations.gtf')\n",
    "check_exons = check_exons[check_exons[\"feature\"]==\"exon\"]\n",
    "check_exons.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "324792\n",
      "1007321\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(datasets_path + 'FinalIntronExonDF.csv')\n",
    "print(len(df[df['feature']==\"intron\"]))\n",
    "print(len(df[df['feature']==\"exon\"]))\n",
    "print(len(df[df['seqname']==\"chrM\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "0.2438171536 % Introns\n",
    "\n",
    "0.7561828464 % Exons\n",
    "\n",
    "3.101434149:1 Exons : Introns\n",
    "\n",
    "357737"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cstart</th>\n",
       "      <th>cend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.007321e+06</td>\n",
       "      <td>1.007321e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.623685e+07</td>\n",
       "      <td>7.623714e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.645594e+07</td>\n",
       "      <td>5.645595e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.071300e+04</td>\n",
       "      <td>1.089200e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.220691e+07</td>\n",
       "      <td>3.220703e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.382008e+07</td>\n",
       "      <td>6.382017e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.124152e+08</td>\n",
       "      <td>1.124159e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489367e+08</td>\n",
       "      <td>2.489371e+08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cstart          cend\n",
       "count  1.007321e+06  1.007321e+06\n",
       "mean   7.623685e+07  7.623714e+07\n",
       "std    5.645594e+07  5.645595e+07\n",
       "min    1.071300e+04  1.089200e+04\n",
       "25%    3.220691e+07  3.220703e+07\n",
       "50%    6.382008e+07  6.382017e+07\n",
       "75%    1.124152e+08  1.124159e+08\n",
       "max    2.489367e+08  2.489371e+08"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_exons.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
